src/persist-types/src/columnar.rs

```diff
@@ -98,7 +98,13 @@ pub trait Data: Debug + Send + Sync + Sized + 'static {
     type Mut: ColumnPush<Self>;
 
     /// The statistics type of columns of this type of data.
-    type Stats: ColumnStats<Self> + for<'a> From<&'a Self::Col>;
+    ///
+    /// The validity, if given, indicates which values in the columns are and
+    /// are not used for stats. This allows us to model non-nullable columns in
+    /// a nullable struct. For optional columns (i.e. ones with their own
```

Comment:
played with this locally. I was hoping we could essentially do `type Col = Buffer<$data>` -> `type Col = PrimitiveArray<$data>`, but realized while typing it up that we actually have to go further and make every columnar type (i.e. the `T::Ref` types) optional so that optional structs can always push `None`. it's very possible that i'm tilting at windmills here, but I really dislike the "everything is optional and non-nullable is emergent" part of arrow/arrow2, and changing the `T::Ref` types to all be Options feels like maybe one step too far down that path?

Reply:
okay, @bkirwi pushed a cleaned up version of what was here the last time you looked at this. the more I look at it, the more I dislike both approaches, so happy to accept pushback on doing it this way vs the other. PTAL!

Reply:
I think my overall feeling is:
- Generally leaning towards sticking to arrow semantics instead of abstracting too heavily. I don't like them either, but they also seem to be pretty hard to abstract over in a way that's both ergonomic and performant, and we seem to need substantial and somewhat painful restructurings of the wrapping code pretty often when we want to add something new.
- Feel like this is probably not the PR to take this stuff in a new direction, and any future changes shouldn't impact the stuff we're writing down, so I'm inclined to stick to what you have here.

Reply:
ack

> and we seem to need substantial and somewhat painful restructurings of the wrapping code pretty often when we want to add something new

fwiw, I view this as more a function of how I tend to develop APIs (iteratively) than anything intrinsic